crates/parsing/src/grammar/rules.rs

```diff
@@ -363,6 +363,7 @@ fn expr_binding(parser: &mut Parser, separator: SyntaxKind) {
         expr_where(parser);
         marker.end(parser, SyntaxKind::UnconditionalBinding);
     } else {
+        // NOTE[et]: We get an infinet loop here since `expr_guarded` needs to beable to match an empty expression, so we just abort this loop if we have looped. It's a crude solution but it works to eliminate infinet loops in the parser. This logic lies inside the `one_or_more` parser
```

Comment:
Do you have an example file that can cause this infinite loop?

Reply:
A minimum reproduction that I conjured is:
```rs
b | if 0 else 1 then 2 = 0
```

This fails due to how `expr_if` is defined:
```rs
fn expr_if(parser: &mut Parser) {
    let mut marker = parser.start();

    parser.expect(SyntaxKind::IfKw);
    expr_0(parser);

    parser.expect(SyntaxKind::ThenKw);
    expr_0(parser);

    parser.expect(SyntaxKind::ElseKw);
    expr_0(parser);

    marker.end(parser, SyntaxKind::IfThenElseExpression);
}
```

More specifically, `expect` is basically `eat` but it emits an error if it isn't able to consume the current token. The new `expect_recover` function is more suitable for this, where it eats the current token indiscriminately while also emitting an error node for the CST.

That being said, loop-safety in the parser is definitely something we should strive for; checking if the parser made progress is a good solution, and I think we should also look into "recovering" as much as possible in the parsing rules.